shared/Microsoft.Extensions.ObjectMethodExecutor.Sources/ObjectMethodExecutor.cs

```diff
@@ -0,0 +1,415 @@
+            MethodInfo awaiterGetResultMethod,
+            MethodInfo awaiterOnCompletedMethod)
+        {
+            // Parameters to executor
+            var targetParameter = Expression.Parameter(typeof(object), "target");
+            var parametersParameter = Expression.Parameter(typeof(object[]), "parameters");
+
+            // Build parameter list
+            var parameters = new List<Expression>();
+            var paramInfos = methodInfo.GetParameters();
+            for (int i = 0; i < paramInfos.Length; i++)
+            {
+                var paramInfo = paramInfos[i];
+                var valueObj = Expression.ArrayIndex(parametersParameter, Expression.Constant(i));
+                var valueCast = Expression.Convert(valueObj, paramInfo.ParameterType);
+
+                // valueCast is "(Ti) parameters[i]"
+                parameters.Add(valueCast);
+            }
+
+            // Call method
+            var instanceCast = Expression.Convert(targetParameter, targetTypeInfo.AsType());
+            var methodCall = Expression.Call(instanceCast, methodInfo, parameters);
+
+            // Using the method return value, construct an ObjectMethodExecutorAwaitable based on
+            // the info we have about its implementation of the awaitable pattern. Note that all
+            // the funcs/actions we construct here are precompiled, so that only one instance of
+            // each is preserved throughout the lifetime of the ObjectMethodExecutor.
+
+            // var getAwaiterFunc = (object awaitable) =>
+            //     (object)((CustomAwaitableType)awaitable).GetAwaiter();
+            var customAwaiterParam = Expression.Parameter(typeof(object), "awaitable");
+            var getAwaiterFunc = Expression.Lambda<Func<object, object>>(
+                Expression.Convert(
+                    Expression.Call(
+                        Expression.Convert(customAwaiterParam, methodInfo.ReturnType),
+                        getAwaiterMethod),
+                    typeof(object)),
+                customAwaiterParam).Compile();
+
+            // var isCompletedFunc = (object awaiter) =>
+            //     ((CustomAwaiterType)awaiter).IsCompleted;
+            var isCompletedParam = Expression.Parameter(typeof(object), "awaiter");
+            var isCompletedFunc = Expression.Lambda<Func<object, bool>>(
+                Expression.MakeMemberAccess(
+                    Expression.Convert(isCompletedParam, awaiterType),
+                    awaiterIsCompletedProperty),
+                isCompletedParam).Compile();
+
+            var getResultParam = Expression.Parameter(typeof(object), "awaiter");
+            Func<object, object> getResultFunc;
+            if (awaiterResultType == typeof(void))
+            {
+                // var getResultFunc = (object awaiter) => (object)null;
+                getResultFunc = Expression.Lambda<Func<object, object>>(
+                    Expression.Constant(null),
```

Comment:
You still need to call `GetResult`, even if it returns void in order to manifest any exception. `GetResult` with throw the exception that the `Task`/awaitable failed with.